File: LC/rough/local_practice/p39.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution
{
public:
    int m, n;
    void solve(vector<vector<char>> &board)
    {
        m = board.size();
        n = board[0].size();

        for (int i = 0; i < m; i++)
        {
            dfs(board, i, 0);
            dfs(board, i, n - 1);
        }
        for (int i = 0; i < n; i++)
        {
            dfs(board, 0, i);
            dfs(board, m - 1, i);
        }

        for (int i = 0; i < m; i++)
        {
            for (int j = 0; j < n; j++)
            {
                if (board[i][j] == 'O')
                {
                    board[i][j] = 'X';
                }
                else if (board[i][j] == '2')
                {
                    board[i][j] = 'O';
                }
            }
        }
    }

    void dfs(vector<vector<char>> &board, int i, int j)
    {
        if (i < 0 || j < 0 || i >= m || j >= n || board[i][j] == 'X')
        {
            return;
        }

        if (board[i][j] == '2')
        {
            return;
        }

        board[i][j] = '2';

        dfs(board, i + 1, j);
        dfs(board, i, j + 1);
        dfs(board, i - 1, j);
        dfs(board, i, j - 1);
    }
};

File: LC/rough/local_practice/p39.py
class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        r = len(heights)
        c = len(heights[0])
        res = []

        pacific = set()
        atlantic = set()

        def dfs(i, j, ocean):
            if (i, j) in ocean:
                return

            ocean.add((i, j))

            dir = [[0, 1], [1, 0], [-1, 0], [0, -1]]

            for dxy in dir:
                nx = i + dxy[0]
                ny = j + dxy[1]

                if nx < 0 or nx >= r or ny < 0 or ny >= c:
                    continue

                if heights[nx][ny] >= heights[i][j]:
                    dfs(nx, ny, ocean)

        for j in range(c):
            dfs(0, j, pacific)
            dfs(r - 1, j, atlantic)

        for i in range(r):
            dfs(i, 0, pacific)
            dfs(i, c - 1, atlantic)

        for i in range(r):
            for j in range(c):
                if (i, j) in pacific and (i, j) in atlantic:
                    res.append([i, j])

        return res
